product_variant_default_code/models/product.py

```diff
@@ -0,0 +1,225 @@
+
+    @api.model
+    def create(self, vals):
+        product = self.new(vals)
+        if (not vals.get('reference_mask') and product.attribute_line_ids or
+                not self.user_has_groups(
+                    'product_variant_default_code.group_product_default_code'
+                )):
+            vals['reference_mask'] = product._get_default_mask()
+        elif vals.get('reference_mask'):
+            sanitize_reference_mask(product, vals['reference_mask'])
+        return super(ProductTemplate, self).create(vals)
+
+    def write(self, vals):
+        product_obj = self.env['product.product']
+        if ('reference_mask' in vals and not vals['reference_mask'] or not
+                self.user_has_groups(
+                    'product_variant_default_code.group_product_default_code'
+                )):
+            if self.attribute_line_ids:
+                vals['reference_mask'] = self._get_default_mask()
+        result = super(ProductTemplate, self).write(vals)
+        if vals.get('reference_mask'):
+            cond = [('product_tmpl_id', '=', self.id),
+                    ('manual_code', '=', False)]
+            products = product_obj.search(cond)
+            for product in products:
+                if product.reference_mask:
+                    render_default_code(product, product.reference_mask)
+        return result
+
+    @api.model
+    def _guess_main_lang(self):
+        """ Used by get_rendered_default_code()
+        """
+        english = self.env.ref('base.lang_en')
+        if english.active:
+            return english.code
+        else:
+            # Naive/simple implementation:
+            # you may inherit to override it in your custom code
+            # to return the language code of your choice
+            return self.env['res.lang'].search([], limit=1).code
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    manual_code = fields.Boolean(string='Manual Reference')
+
+    @api.model
+    def create(self, vals):
+        product = super(ProductProduct, self).create(vals)
+        if product.reference_mask:
+            render_default_code(product, product.reference_mask)
+        return product
+
+    @api.onchange('default_code')
+    def onchange_default_code(self):
+        self.manual_code = bool(self.default_code)
+
+
+class ProductAttribute(models.Model):
+    _inherit = 'product.attribute'
+
+    code = fields.Char(
+        string='Attribute Code',
+    )
+
+    _sql_constraints = [
+        ('number_uniq', 'unique(name)', _('Attribute Name must be unique!'))]
+
+    def write(self, vals):
+        if 'code' not in vals:
+            return super(ProductAttribute, self).write(vals)
+        result = super(ProductAttribute, self).write(vals)
+        # Rewrite reference on all product variants affected
+        for product in self.mapped('attribute_line_ids').mapped(
+            'product_tmpl_id').mapped('product_variant_ids').filtered(
```

Comment:
why use a chain of mapped() instead of mapped('attribute_line_ids.product_tmpl_id.product_variant_ids') ?